src/main/scala/com/databricks/spark/csv/util/InferSchema.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2014 Databricks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.spark.csv.util
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.types._
+
+import scala.util.control.Exception._
+
+private[csv] object InferSchema {
+
+  /**
+   * Similar to the JSON schema inference. {@link org.apache.spark.sql.json.InferSchema}
+   *     1. Infer type of each row
+   *     2. Merge row types to find common type
+   *     3. Replace any null types with string type
+   */
+  def apply(tokenRdd: RDD[Array[String]], header: Array[String]): StructType = {
+
+    val rootTypes = tokenRdd.aggregate(Array[DataType]())(inferRowType, mergeRowTypes)
+
+    val stuctFields = header.zip(rootTypes).map { case (thisHeader, rootType) =>
+      StructField(thisHeader, rootType, nullable = true)
+    }
+
+    StructType(stuctFields)
+  }
+
+  private[csv] def inferRowType(rowSoFar: Array[DataType], next: Array[String]) = {
+    rowSoFar.zipAll(next, NullType, null).map{case (tpe, field) => inferField(tpe, field)}
```

Comment:
the inferField enhancement looks ok...however i am not sure if inferRowType() will avoid the GC issue as zipAll and map will create temp objects. rdd.aggregate() allows you to have a "mutable" left param...so you could do something like this:

var i = 0
while(i < rowSoFar.length) {
 rowSoFar(i) = inferField(rowSoFar(i), field) // not sure if findTightestCommonType needs  be called here?
}